pylint/checkers/similar.py

```diff
@@ -43,6 +68,276 @@
+        "_hash",  # The hash of some consecutive lines
+    )
+
+    def __init__(self, fileid: str, num_line: int, *lines: Iterable[str]):
+        self._fileid: str = fileid
+        self._index: Index = Index(num_line)
+        self._hash: int = 0
+        for lin in lines:
+            if lin:
+                self._hash += hash(lin)
+            else:
+                # Avoid matching empty lines
+                self._hash += hash(int(random.random() * 100000))
+
+    def __eq__(self, o: object) -> bool:
+        if not isinstance(o, LinesChunk):
+            return NotImplemented
+        return self._hash == o._hash
+
+    def __hash__(self) -> int:
+        return self._hash
+
+    def __repr__(self) -> str:
+        return (
+            f"<LinesChunk object for file {self._fileid} ({self._index}, {self._hash})>"
+        )
+
+    def __str__(self) -> str:
+        return (
+            f"LinesChunk object for file {self._fileid}, starting at line {self._index} \n"
+            f"Hash is {self._hash}"
+        )
+
+
+class SuccessiveLinesLimits:
+    """
+    A class to handle the numbering of begin and end of successive lines.
+
+    :note: Only the end line number can be updated.
+    """
+
+    def __init__(self, start: LineNumber, end: LineNumber):
+        self._start: LineNumber = start
+        self._end: LineNumber = end
+
+    @property
+    def start(self) -> LineNumber:
+        return self._start
+
+    @property
+    def end(self) -> LineNumber:
+        return self._end
+
+    @end.setter
+    def end(self, value: LineNumber):
+        self._end = value
+
+    def __repr__(self) -> str:
+        return f"<SuccessiveLinesLimits <{self._start};{self._end}>>"
+
+
+class LineSetStartCouple:
+    """
+    Indices in both linesets that mark the beginning of successive lines
+    """
+
+    def __init__(self, fst_lineset_index: Index, snd_lineset_index: Index):
+        self._fst_lineset_index = fst_lineset_index
+        self._snd_lineset_index = snd_lineset_index
+
+    def __repr__(self) -> str:
+        return f"<LineSetStartCouple <{self._fst_lineset_index};{self._snd_lineset_index}>>"
+
+    def __add__(self, value: Index):
+        return LineSetStartCouple(
+            Index(self._fst_lineset_index + value),
+            Index(self._snd_lineset_index + value),
+        )
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, LineSetStartCouple):
+            raise NotImplementedError
+        return (
+            self._fst_lineset_index == other._fst_lineset_index
+            and self._snd_lineset_index == other._snd_lineset_index
+        )
+
+    def __hash__(self) -> int:
+        return hash(self._fst_lineset_index) + hash(self._snd_lineset_index)
+
+    @property
+    def fst_lineset_index(self) -> Index:
+        return self._fst_lineset_index
+
+    @property
+    def snd_lineset_index(self) -> Index:
+        return self._snd_lineset_index
```

Comment:
I'm not really sure how much it will be faster but its definitely clearer, shorter and smarter. Thanks @Pierre-Sassoulas .

Reply:
In fact inheriting from `NamedTuple` prevents the definition of `__add__` method as i did it because then it doesn't respect the Liskov Substitution Principle as pointed out by `mypy` and https://stackoverflow.com/questions/65498173/mypy-signature-of-add-incompatible-with-supertype-tuple-but-everythin.
I changed the name of `__add__` method in increment which is IMHO better suited.